{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0c6e4d88-15ef-4f73-9223-9c7dd428e40d",
   "metadata": {},
   "source": [
    "# Simulated Data for Brian Kirz's CS-282 Final Project\n",
    "First I load the necssary packages and double check key packages versions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ba3b4126-c56c-4e39-8f0a-8dad77d8801d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "demes 0.2.1\n",
      "demesdraw 0.3.0\n",
      "msprime 1.1.1\n",
      "tskit 0.4.1\n"
     ]
    }
   ],
   "source": [
    "# Import packages.\n",
    "import demes\n",
    "import demesdraw\n",
    "from IPython.display import display, SVG\n",
    "from matplotlib import pyplot as plt\n",
    "import msprime\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import tskit\n",
    "# Print versions of our libraries.\n",
    "print('demes', demes.__version__)\n",
    "print('demesdraw', demesdraw.__version__)\n",
    "print('msprime', msprime.__version__)\n",
    "print('tskit', tskit.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "916045e6-2082-4ceb-8e0a-57f01cfdf916",
   "metadata": {},
   "source": [
    "If the output of the above cell doesn't match `demes 0.2.1`, `demesdraw 0.3.0`, `msprime 1.1.1`, and `tskit 0.4.1` then this notebook will throw you one million error messages. I will now define two general functions, that I will need for the rest of this notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35c8f5ab-7358-4b46-b5f2-b3b635a5fc46",
   "metadata": {},
   "source": [
    "## Model Sanity Check\n",
    "Next I will use the same model as described in Fang et al. 2022 and validate it using `msprime` and `demes`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "882acb15-a8f8-4d9a-9444-ee487448d542",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><details open=\"true\"><summary>Epoch[0]: [0, 4e+03) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=7 active=4)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>AFR</th><th>EUR</th><th>NEA</th><th>OUT</th></tr></thead><tbody><tr><td>AFR</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population AFR to EUR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from EUR to AFR per generation, divided by the size of AFR.'>0</td><td title='Backwards in time migration rate from population AFR to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to AFR per generation, divided by the size of AFR.'>0</td><td title='Backwards in time migration rate from population AFR to OUT = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OUT to AFR per generation, divided by the size of AFR.'>0</td></tr><tr><td>EUR</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population EUR to AFR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AFR to EUR per generation, divided by the size of EUR.'>0</td><td>0</td><td title='Backwards in time migration rate from population EUR to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to EUR per generation, divided by the size of EUR.'>0</td><td title='Backwards in time migration rate from population EUR to OUT = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OUT to EUR per generation, divided by the size of EUR.'>0</td></tr><tr><td>NEA</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population NEA to AFR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AFR to NEA per generation, divided by the size of NEA.'>0</td><td title='Backwards in time migration rate from population NEA to EUR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from EUR to NEA per generation, divided by the size of NEA.'>0</td><td>0</td><td title='Backwards in time migration rate from population NEA to OUT = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OUT to NEA per generation, divided by the size of NEA.'>0</td></tr><tr><td>OUT</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population OUT to AFR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AFR to OUT per generation, divided by the size of OUT.'>0</td><td title='Backwards in time migration rate from population OUT to EUR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from EUR to OUT per generation, divided by the size of OUT.'>0</td><td title='Backwards in time migration rate from population OUT to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to OUT per generation, divided by the size of OUT.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events @ generation 4e+03</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>4000</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_mass_migration'>Mass Migration</a></td><td>source=EUR, dest=NEA, proportion=0.03</td><td>Lineages currently in population EUR move to NEA with probability 0.03 (equivalent to individuals migrating from NEA to EUR forwards in time)</td></tr></tbody></table></div></div></details><details open=\"true\"><summary>Epoch[1]: [4e+03, 1.6e+04) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=7 active=4)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>AFR</th><th>EUR</th><th>NEA</th><th>OUT</th></tr></thead><tbody><tr><td>AFR</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population AFR to EUR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from EUR to AFR per generation, divided by the size of AFR.'>0</td><td title='Backwards in time migration rate from population AFR to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to AFR per generation, divided by the size of AFR.'>0</td><td title='Backwards in time migration rate from population AFR to OUT = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OUT to AFR per generation, divided by the size of AFR.'>0</td></tr><tr><td>EUR</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population EUR to AFR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AFR to EUR per generation, divided by the size of EUR.'>0</td><td>0</td><td title='Backwards in time migration rate from population EUR to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to EUR per generation, divided by the size of EUR.'>0</td><td title='Backwards in time migration rate from population EUR to OUT = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OUT to EUR per generation, divided by the size of EUR.'>0</td></tr><tr><td>NEA</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population NEA to AFR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AFR to NEA per generation, divided by the size of NEA.'>0</td><td title='Backwards in time migration rate from population NEA to EUR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from EUR to NEA per generation, divided by the size of NEA.'>0</td><td>0</td><td title='Backwards in time migration rate from population NEA to OUT = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OUT to NEA per generation, divided by the size of NEA.'>0</td></tr><tr><td>OUT</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population OUT to AFR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AFR to OUT per generation, divided by the size of OUT.'>0</td><td title='Backwards in time migration rate from population OUT to EUR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from EUR to OUT per generation, divided by the size of OUT.'>0</td><td title='Backwards in time migration rate from population OUT to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to OUT per generation, divided by the size of OUT.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events @ generation 1.6e+04</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>1.6e+04</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[AFR, EUR], ancestral=AMH</td><td>Moves all lineages from derived populations &#x27;AFR&#x27; and &#x27;EUR&#x27; to the ancestral &#x27;AMH&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open=\"true\"><summary>Epoch[2]: [1.6e+04, 2e+04) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=7 active=3)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>NEA</th><th>OUT</th><th>AMH</th></tr></thead><tbody><tr><td>NEA</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population NEA to OUT = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OUT to NEA per generation, divided by the size of NEA.'>0</td><td title='Backwards in time migration rate from population NEA to AMH = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AMH to NEA per generation, divided by the size of NEA.'>0</td></tr><tr><td>OUT</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population OUT to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to OUT per generation, divided by the size of OUT.'>0</td><td>0</td><td title='Backwards in time migration rate from population OUT to AMH = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AMH to OUT per generation, divided by the size of OUT.'>0</td></tr><tr><td>AMH</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population AMH to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to AMH per generation, divided by the size of AMH.'>0</td><td title='Backwards in time migration rate from population AMH to OUT = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OUT to AMH per generation, divided by the size of AMH.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events @ generation 2e+04</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>2e+04</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[AMH, NEA], ancestral=HUM</td><td>Moves all lineages from derived populations &#x27;AMH&#x27; and &#x27;NEA&#x27; to the ancestral &#x27;HUM&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open=\"true\"><summary>Epoch[3]: [2e+04, 8e+05) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=7 active=2)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>OUT</th><th>HUM</th></tr></thead><tbody><tr><td>OUT</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population OUT to HUM = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from HUM to OUT per generation, divided by the size of OUT.'>0</td></tr><tr><td>HUM</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population HUM to OUT = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from OUT to HUM per generation, divided by the size of HUM.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events @ generation 8e+05</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>8e+05</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[HUM, OUT], ancestral=ANC</td><td>Moves all lineages from derived populations &#x27;HUM&#x27; and &#x27;OUT&#x27; to the ancestral &#x27;ANC&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open=\"true\"><summary>Epoch[4]: [8e+05, inf) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=7 active=1)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>ANC</td><td> 10000.0</td><td> 10000.0</td><td> 0</td></tr></tbody></table></div></div></details></div>"
      ],
      "text/plain": [
       "<msprime.demography.DemographyDebugger at 0x7fc356e6c6d0>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define IUA model of introgression.\n",
    "def iua_human_model(f):\n",
    "    # Intialize demographic model.\n",
    "    iua_model = msprime.Demography()\n",
    "    # We assume constant and equal effective population sizes for\n",
    "    # all lineages.\n",
    "    iua_model.add_population(name='AFR', initial_size=10_000)\n",
    "    iua_model.add_population(name='EUR', initial_size=10_000)\n",
    "    iua_model.add_population(name='NEA', initial_size=10_000)\n",
    "    iua_model.add_population(name='OUT', initial_size=10_000)\n",
    "    iua_model.add_population(name='AMH', initial_size=10_000)\n",
    "    iua_model.add_population(name='HUM', initial_size=10_000)\n",
    "    iua_model.add_population(name='ANC', initial_size=10_000)\n",
    "    # Introgression from the Neanderthal to the Eurasian lineage\n",
    "    # occuring 4,000 generations ago with a probability of f.\n",
    "    iua_model.add_mass_migration(\n",
    "        time=4_000, source='EUR', dest='NEA', proportion=f,\n",
    "    )\n",
    "    # The African and Eurasian lineages merge into the anatomically\n",
    "    # modern human lineage 16,000 generations ago.\n",
    "    iua_model.add_population_split(\n",
    "        time=16_000, derived=['AFR', 'EUR'], ancestral='AMH',\n",
    "    )\n",
    "    # The anatomically modern human and Neanderthal lineages merge\n",
    "    # into the ancestral human lineage 20,000 generations ago.\n",
    "    iua_model.add_population_split(\n",
    "        time=20_000, derived=['AMH', 'NEA'], ancestral='HUM',\n",
    "    )\n",
    "    # The ancestral human and outgroup lineages merge\n",
    "    # into the ancestral lineage 800,000 generations ago.\n",
    "    iua_model.add_population_split(\n",
    "        time=800_000, derived=['HUM', 'OUT'], ancestral='ANC',\n",
    "    )\n",
    "    return iua_model\n",
    "# Sanity check to make sure the model is right.\n",
    "test_model = iua_human_model(0.03)\n",
    "test_model.debug()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e961a189-46a2-4090-a2c8-dee635b2d86a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><details open=\"true\"><summary>Epoch[0]: [0, 4e+03) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=5 active=3)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>AFR</th><th>EUR</th><th>NEA</th></tr></thead><tbody><tr><td>AFR</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population AFR to EUR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from EUR to AFR per generation, divided by the size of AFR.'>0</td><td title='Backwards in time migration rate from population AFR to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to AFR per generation, divided by the size of AFR.'>0</td></tr><tr><td>EUR</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population EUR to AFR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AFR to EUR per generation, divided by the size of EUR.'>0</td><td>0</td><td title='Backwards in time migration rate from population EUR to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to EUR per generation, divided by the size of EUR.'>0</td></tr><tr><td>NEA</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population NEA to AFR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AFR to NEA per generation, divided by the size of NEA.'>0</td><td title='Backwards in time migration rate from population NEA to EUR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from EUR to NEA per generation, divided by the size of NEA.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events @ generation 4e+03</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>4000</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_mass_migration'>Mass Migration</a></td><td>source=EUR, dest=NEA, proportion=0.03</td><td>Lineages currently in population EUR move to NEA with probability 0.03 (equivalent to individuals migrating from NEA to EUR forwards in time)</td></tr></tbody></table></div></div></details><details open=\"true\"><summary>Epoch[1]: [4e+03, 1.6e+04) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=5 active=3)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>AFR</th><th>EUR</th><th>NEA</th></tr></thead><tbody><tr><td>AFR</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population AFR to EUR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from EUR to AFR per generation, divided by the size of AFR.'>0</td><td title='Backwards in time migration rate from population AFR to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to AFR per generation, divided by the size of AFR.'>0</td></tr><tr><td>EUR</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population EUR to AFR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AFR to EUR per generation, divided by the size of EUR.'>0</td><td>0</td><td title='Backwards in time migration rate from population EUR to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to EUR per generation, divided by the size of EUR.'>0</td></tr><tr><td>NEA</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population NEA to AFR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AFR to NEA per generation, divided by the size of NEA.'>0</td><td title='Backwards in time migration rate from population NEA to EUR = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from EUR to NEA per generation, divided by the size of NEA.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events @ generation 1.6e+04</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>1.6e+04</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[AFR, EUR], ancestral=AMH</td><td>Moves all lineages from derived populations &#x27;AFR&#x27; and &#x27;EUR&#x27; to the ancestral &#x27;AMH&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open=\"true\"><summary>Epoch[2]: [1.6e+04, 2e+04) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=5 active=2)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th><th>NEA</th><th>AMH</th></tr></thead><tbody><tr><td>NEA</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td>0</td><td title='Backwards in time migration rate from population NEA to AMH = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from AMH to NEA per generation, divided by the size of NEA.'>0</td></tr><tr><td>AMH</td><td> 10000.0</td><td> 10000.0</td><td> 0</td><td title='Backwards in time migration rate from population AMH to NEA = 0.0 per\n",
       "generation. Forwards in time, this is the expected number of migrants moving\n",
       "from NEA to AMH per generation, divided by the size of AMH.'>0</td><td>0</td></tr></tbody></table></div><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Events @ generation 2e+04</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th>time</th><th>type</th><th>parameters</th><th>effect</th></tr></thead><tbody><tr><td>2e+04</td><td><a href='https://tskit.dev/msprime/docs/latest/api.html#msprime.Demography.add_population_split'>Population Split</a></td><td>derived=[AMH, NEA], ancestral=HUM</td><td>Moves all lineages from derived populations &#x27;AMH&#x27; and &#x27;NEA&#x27; to the ancestral &#x27;HUM&#x27; population. Also set the derived populations to inactive, and all migration rates to and from the derived populations to zero.</td></tr></tbody></table></div></div></details><details open=\"true\"><summary>Epoch[3]: [2e+04, inf) generations</summary><div style=\"margin-left:20px\"><div><style scoped=\"\">\n",
       "            .tskit-table thead tr th:only-of-type {vertical-align: middle;}\n",
       "            .tskit-table thead tr th {text-align: center;vertical-align: top;}\n",
       "            .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "            .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "        </style><b>Populations (total=5 active=1)</b><table border=\"1\" class=\"tskit-table\"><thead><tr><th></th><th>start</th><th>end</th><th>growth_rate</th></tr></thead><tbody><tr><td>HUM</td><td> 10000.0</td><td> 10000.0</td><td> 0</td></tr></tbody></table></div></div></details></div>"
      ],
      "text/plain": [
       "<msprime.demography.DemographyDebugger at 0x7fc342fd7610>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define IUA model of introgression.\n",
    "def iua_human_model(f):\n",
    "    # Intialize demographic model.\n",
    "    iua_model = msprime.Demography()\n",
    "    # We assume constant and equal effective population sizes for\n",
    "    # all lineages.\n",
    "    iua_model.add_population(name='AFR', initial_size=10_000)\n",
    "    iua_model.add_population(name='EUR', initial_size=10_000)\n",
    "    iua_model.add_population(name='NEA', initial_size=10_000)\n",
    "    iua_model.add_population(name='AMH', initial_size=10_000)\n",
    "    iua_model.add_population(name='HUM', initial_size=10_000)\n",
    "    # Introgression from the Neanderthal to the Eurasian lineage\n",
    "    # occuring 4,000 generations ago with a probability of f.\n",
    "    iua_model.add_mass_migration(\n",
    "        time=4_000, source='EUR', dest='NEA', proportion=f,\n",
    "    )\n",
    "    # The African and Eurasian lineages merge into the anatomically\n",
    "    # modern human lineage 16,000 generations ago.\n",
    "    iua_model.add_population_split(\n",
    "        time=16_000, derived=['AFR', 'EUR'], ancestral='AMH',\n",
    "    )\n",
    "    # The anatomically modern human and Neanderthal lineages merge\n",
    "    # into the ancestral human lineage 20,000 generations ago.\n",
    "    iua_model.add_population_split(\n",
    "        time=20_000, derived=['AMH', 'NEA'], ancestral='HUM',\n",
    "    )\n",
    "    return iua_model\n",
    "# Sanity check to make sure the model is right.\n",
    "test_model = iua_human_model(0.03)\n",
    "test_model.debug()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1524570d-c1e0-418c-afb4-ed0050c11b3c",
   "metadata": {},
   "source": [
    "Looks good but lets visualize this model as a sanity check."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1502dd5a-097f-4de0-915d-ecbca0ef4d53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='time ago (generations)'>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtMAAAIFCAYAAAAQk3EDAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAArrElEQVR4nO3debikZX0n/O+vz9IbcNgVWWyklSXGUWwxJmZTRDC2oskYGDXRGDqOa0wyURMzmuRNTCYxk3FCdPCKS3wjDklcaERxeWXUxLxiXCKLKCEYGgVka2h6O919zx91wKbppXi6q6uq+/O5rrr6PHdVPfU9TVPne566n/up1loAAICHbt6wAwAAwLhSpgEAoCNlGgAAOlKmAQCgI2UaAAA6UqYBAKCjyWEH2B1nnnlm+8QnPjHsGDDyfv2ir+Xnnnhs/un624cdhT3kRx51WP7p+tvzumc8ZthRAPYXtb3BsT4yfdtttw07AoyFv//KTXnKCYcNOwZ70FNOOCz/4zPfHnYMgP3eWJdpAAAYJmUaAAA6UqYBAKAjZRoAADpSpgEAoKOxLNNVtbyqLli9evWwowAAsB8byzLdWlvZWlsxMzMz7CgAAOzHxrJMAwDAKFCmAQCgI2UaAAA6UqYBAKAjZRoAADpSpgEAoCNlGgAAOlKmAQCgI2UaAAA6UqYBAKAjZRoAADpSpgEAoCNlGgAAOlKmAQCgI2UaAAA6UqYBAKAjZRoAADpSpgEAoCNlGgAAOlKmAQCgI2UaAAA6GpkyXVU/VVWfr6p3VtVPDTsPAADsykDLdFW9u6puraortxk/s6qurarrquoNc8MtyZokC5KsGmQuAADYEwZ9ZPq9Sc7ceqCqJpKcn+SsJKckObeqTkny+dbaWUlen+R3B5wLAAB220DLdGvtc0nu2Gb4tCTXtdaub61tTPLBJM9trW2Zu//OJPMHmQsAAPaEySG85tFJbtxqe1WSJ1fV85M8M8nBSf5iR0+uqhVJViTJcccdN7iUAACwC8Mo07WdsdZa+1CSD+3qya21C5JckCTLli1rezgbAAD0bRireaxKcuxW28ck+e4QcgAAwG4ZRpm+Ismjq+r4qppOck6Si4eQAwAAdsugl8a7MMkXk5xYVauq6mWttU1JXpXksiTXJLmotXbVQ9zv8qq6YPXq1Xs+NAAA9Gmgc6Zba+fuYPzSJJfuxn5XJlm5bNmy87ruAwAAdtfIXAERAADGjTINAAAdDWNpPIBdmlk4lccdM5PpycH+zn/X2tl87ca7snmLlTYBeOjG8si0ExBh33fyUQcNvEgnycGLpvLIwxYN/HUA2DeNZZlura1sra2YmZkZdhRgQKYmfnB9px858ZgH3PfRiz6QP3zTf0mS/M7rXpFPfeyjD7j/vsffdOO/5z8ce0j+4k/+4P777rzj9jzx+CPuf37vtcbyrRCAEeAnCLBPO+a4JfncZy67f/tTl3wkJzzmpCEmAmBfokwD+7T5CxbkUUsfk6u+/tUkyWUrP5wznn32cEMBsM9wAiIw8jasX5cXPPPH799efded+clnnNX38898zvPziYs/lMOPPDLzJiZyxMOOyq233DyIqADsZ8ayTFfV8iTLly5dOuwowF4wf8HCXHTZ5+/f/uhFH8hV/9I70pyqBz2+thn7sZ86Pef/6R/msCOOyDOXP2+gWQHYv4zlNA8nIAL3OfiQQ3P36rvu31595505+NBDH/CYqenpnPzDj89fX3B+Tn/Wc/ZyQgD2ZWNZpgHus+wpP5bLVn44sxs3Jkk++rcfyJOe8uMPetwvrHhlfvWNb8nBhxz6oPsAoKuxnOYBcJ+fPP3MXPONr+ecZ/1UJiYmcswjj8+b3vpnD3rc0hNPztITTx5CQgD2ZdXa+F71a9myZe3LX/7ysGPAyFvyho/lhj/6mfz3T31r2FH69uOPPnyvXLQlSVbduS7X3nzPXnmtPeV1z3jM/f9dAdgrHnySTkzzAACAzsayTLucOOz7ZjfvvU/NZjdt2WuvBcC+ZSzLtNU8YN939XdXZ/3slrSWgd22tOSOezfmO3esHfa3C8CYcgIiMJLuXr8p/3DdbcOOAQA7NZZHpgEAYBQo0wAA0JEyDQAAHSnTAADQkTINAAAdjWWZts40AACjYCzLtHWmAQAYBWNZpgEAYBQo0wAA0JEyDQAAHSnTAADQkTINAAAdKdMAANCRMg0AAB2NZZl20RYAAEbBWJZpF20BAGAUjGWZBgCAUaBMAwBAR8o0AAB0pEwDAEBHyjQAAHSkTAMAQEfKNAAAdKRMAwBAR8o0AAB0NJZl2uXEAQAYBWNZpl1OHACAUTCWZRoAAEaBMg0AAB0p0wAA0JEyDQAAHSnTAADQkTINAAAdKdMAANCRMg0AAB0p0wAA0JEyDQAAHSnTAADQkTINuzC7eUtW3bl22DFgn9Fayy+/74pc+KV/H3YUgN02OewAMKru3bAp//uKG/N7l1ydJLnhj37mQY/ZsqVl3ezmrNmwKfesn826jVtyz4bZrN2wOetme7f1c/ev27g5GzdvyabNLbObt2RLa2mtt595VZmcqExPzsv8yYksnJrIwql5mZqcl+mJeVkwNZHF8ydy0IKpHLxoOgctmMyBC6ayYGpeqmpv/rUwQo6aWTDsCJ19+ppb8+lrbs11t67J657xmBww348jYDx594Jt3Hr3+pz2h59Jkjzrhx+eJLnk1U/Nh76yKp+99vu5fc2G3LZmQ25bszF3rd2YBVMTWTQ9mYMWTmbh1EQWz5/MgfMns2BqIgumJrJwel4Wz5/MoqnJLJyeyNS8eZmaqMybV70i3Fq2tN4R8I2bt2Tjpi25a+3GfG92c2Y3b8mGTVuyYXZL7t24KavXzeautbO5Z/2m3L1+NmnJovkTP3jduZJ98MJe0f75Jx2bJz7y0GH+dTJAv3/2Y/PXX7wh13//3iycnsihi6azeH7v3+Li6cnMn5qXA+dPZWbhVA5Y0Pv3MTUx2A8k7/sF8+71s1m7cXPuWb8pt969PnfcuzG33L0h96yfzdNPPjK//ozH5Ka71uWmO9flsW++LK952tL82hknDjTb3vbhD384z3/+83PNNdfkpJNOyg033JDjjz8+b3rTm/L7v//7SZLbbrstRx11VH7lV34lf/EXf5G3vOUtOeCAA/Ibv/Eb9+9nyZIl+fKXv5zDDz98WN8KsBNjWaaranmS5UuXLh1OgA+/PHn8C5MbvjCc12fPW/LU3n/Pn35jZhZO5aU/tiTv+YcbeiVk/mQWz5/Mgql5Of3kI3PIoukcceD8HHbAdA5bPD8T84Z3ZHj97Oas3dg7Ar5m/aas2bApd6+bzep1s1k/uzmHHzA/SfLapz86SfK6ZzxmaFnZ846aWZCpiXmZ3dyyfnZzvrt6XdZt7BXZezf84N/F6nWzWbNhU+7dsCkLpydy4PzeL3bzJycyPTkv05Pzsmh6Igu22p6875e99KZlbN7Sep+sbGnZuGlL1qzfdP8nL+tmN2fdXHFev2lz5k/Oy8zCqSyenswBCyZz5IHzc+ji6Rx54IIcedD8LFtyaP7qCzfk09fckoMXTeXYQxdm7cbND/jefvsLv52zl56dK26+Yhh/tQ/Z9MR0nnLUUx4w9r/e+79y6pNPzdvf/fa88jdfmZvuuCnHLDkmf/eRv8s5rz0nSfLB93wwJ5x4Qu5Yd0euuu2q3Lr21qypNbnqtqvu38/sltl88/Zv5pAckiTZsHlD/ul7/7T3vrk94EkPf1KuuPmKvOLxrxh2FNjjxrJMt9ZWJlm5bNmy84YS4OsXJs97pzK9Lzn+x5P3PTv56Tdm/tRE3rz8h/K6Zzwmf/6pb2fNhk354BX/njeedfKwUz7IfUe/d0WJ3jf90CNmkiQ/+Zgj+nr8li0td6/vfbKxfnZz71OPTVsyu3lL1m7clPWzvU9GNm7aks1zBTrpTUOamJdMTczL5ERv6tEBc4V84dREr4hPTeTABb1PZ+bt4hfM29dsyKevuSWnn3xkfve5j83RBy980GMu/teL8wdP/YOxKdPbWrtmbb76pa/m3R95d179olfnlb/5yiTJggUL8qjHPCpXfu3KPPbxj80nPvKJPPO5z8z3b/7+kBMP1pMe/qT80mW/pEyzTxrLMg17w0ELpvJfl5+SJYcvyuOOOXjYcWC3zZtXOXjRdA5eND3UHIcdMD//9tZn7dPz/T/z8c/kx572Y1lywpIcdMhBufrrV2fmkN4vP2c976x8/MMfz+FHHp6JiYkc+fAjH1Cm3//O9+eSv73k/u1bb751r+cH+qdMwy78wlOWDDsC7HP25SKdJJd+6NK8+FdenCQ56+yzcumHL825v3RukuSpT3tq/ucf/c8cdsRhOfPsMx/03Be//MV56Stfev/2GaeesXdCA50o0wCwB911x1350he+lOu+eV2qKps3b05V5ZyX9uZJT01P5ZTHnZL3/eX78tEvfDSXX3b5cAMDu0WZBoA96JMrP5nnvOA5efPb3nz/2Eue85Lc8t1bfrD9ipfkST/6pBx86MFDSAjsSS7aAkAnt6+7PZfdcNmwY4ycSz90aZ7+rKc/YOz0Z5+ed/35u+7fXnrS0jz3nOfu7Wj7nNP+5rS85v97Tdp9i/bDEDgyDUBf1s2uy9TEVL537/fy/qvfnwu/eWGS5IcP/+Gc/7XzMz0xnel505k/MT8LJhdk0eSiHLLgkBw4fWAOmDogB80/KAdMHZAFkwuycHJhFk0u2ifnTr/3o+990NiLVrwoL1rxou0+/uxzz87Z556dJPev+rG1T37lk3sy3j7luAOPy2dv/Gx+6bJfypt+5E054eAThh2J/ZAyDUB/KnnC+5+QJPnFU34xJ8yckN887Tdz98a7s+xhyzK7ZTYbN2/Mhs0bsn7z+ty67tZ8685v5Z6N92TN7JrcvfHurNm4Jus3r8+6TeuycfPGLJxcmEMWHJJFk4ty8IKD887T35nJeX407YtOPOTEfOfu7+TN//jmTM6bzIFTB2Zm/kwWTS3KgdMHZuHEwiycXJgDpg/IQdMHZfHU4szMn+k9ZnJRFkwuyLx64AfqZx1/Vp581JOzdtPanP3Rs7PsYcvynjPfM6TvkP2Vdyxgj/r0e6/OyU85Kjd9685hR2EPOPoxh+Smb92Z05Y/KgsnF+ZvnvU3eftX3p5//N4/5l9X/2sWTi7MSYeelJMOPekh73t2y2zWbVqXO9ffmbWzazMxb+L+Iv2f/8N/ThLrEu/AqQ87ddgRHrIzlpyRYw44Jq899bVZt2ld1mxck7s23JW1s2vv/0Vr3aZ1uWfjPff/ArZ6w+rcvfHurJ1dm/Wb12dy3mSm5k1l4cTCfOz5H8uF37wwt6y9JUcsPCIvPPmFOeWwUx7wmje94Y05+HnPy9ovfWlI3zV70qLTTsvaL30pR7z6VcOO8gDKNLBHXftPN+f0l5yiTO8jjj7xkHzkv381py1/VJLkcUc8Lu8641351Hc+lV//P7+eN//jm3Px2Rd32vfUvKlMTU/loOmDHnSfEr3vWfG4FUmSJxz5hE7Pb63d/+nHxs0bs7ltzi1rb8mvnvqrefEpL869s/dm4+aND3jO3R/5SI7+o7cq0/uIxU8+Lf/+i7+oTAMw3qoqZyw5I5ccekkmatdX4IQ9oap68/InfnDRoW/84jfu/3rrcdiblGkAOnnkQY8cdgSAobM0HgAAdKRMAwBAR8o0AAB0pEwDAEBHyjQAAHSkTAMAQEfKNAAAdKRMAwBAR8o0AAB0pEwDAEBHI1Wmq2pxVf1zVT172FkAAGBXBlqmq+rdVXVrVV25zfiZVXVtVV1XVW/Y6q7XJ7lokJkAAGBPGfSR6fcmOXPrgaqaSHJ+krOSnJLk3Ko6papOT3J1klsGnAkAAPaIyUHuvLX2uapass3waUmua61dnyRV9cEkz01yQJLF6RXsdVV1aWttyyDzAQDA7hhomd6Bo5PcuNX2qiRPbq29Kkmq6iVJbttRka6qFUlWJMlxxx032KQAALATwzgBsbYz1u7/orX3ttYu2dGTW2sXtNaWtdaWHXHEEQMJCAAA/RhGmV6V5Nitto9J8t0h5AAAgN0yjDJ9RZJHV9XxVTWd5JwkFw8hBwAA7JZBL413YZIvJjmxqlZV1ctaa5uSvCrJZUmuSXJRa+2qh7jf5VV1werVq/d8aAAA6NOgV/M4dwfjlya5dDf2uzLJymXLlp3XdR8AALC7RuoKiAAAME6UaQAA6EiZBgCAjsayTDsBEQCAUTCWZbq1trK1tmJmZmbYUQAA2I+NZZkGAIBRoEwDAEBHyjQAAHQ0lmXaCYgAAIyCsSzTTkAEAGAUjGWZBgCAUaBMAwBAR8o0AAB0pEwDAEBHY1mmreYBAMAoGMsybTUPAABGwViWaQAAGAXKNAAAdKRMAwBAR8o0AAB0pEwDAEBHyjQAAHQ0lmXaOtMAAIyCsSzT1pkGAGAUjGWZBgCAUaBMAwBAR8o0AAB0pEwDAEBHyjQAAHSkTAP9+4e3J+96evKdfxx2EgAYCZPDDgAMyKYNSWvJ1ILk3z6frLklWXtHMv+A5PH/KfmXi5JvXZasvyuZXZ+84H3Jbd9OPvbryab1yeza5Gm/kzzhhcnvH5Fs3viDfb/nrOTXvpkcdNTQvj0AGAVjWaaranmS5UuXLh12FPYXn31r8oU/S6YWJYsPT875QFITyad+J5lamMw/MDn5OcnSpydf+0AyuaA3tviI5BGPT1bflGxck0wv7t238NDeftuWZN7EA19rwz3Jlk29MlyVHPjw5OYrkzU390rv7NrkcS9IbrwiufZjyca1yfrVyVN/NZmYTt79zGT93UnbnDz1dcnT3tTLtGldsuiw5ODjeq8zc2zy6DOShYf0Cvf0AcnDfih5/v9KJuYn04t6j0+SN97Uy/kPf55c9eHkjP+nlwsA9nNjWaZbayuTrFy2bNl5w87CfuInX98rqxvXJmtv6xXSTeuTx78wmV2XbLg7WTB3EaGbr0xW35hsvDc5ZEmvTH/5r5KrP9ob27QhefU/Jzd8Pvnbl/QKdZKc/Y7ksT+b/NnJveI6uaBX0H/mT3tl+NarkqnFyaJDemU6rVfOFx/Ze+2FhyaLDk1e/g+9Ij+1sFfGk+R573jw9/TIp/RuW5takDz8hx/82Mnp3p8//uu9GwCQZEzLNOx18+Yl8xb2CuriuaO1UwuTk5/94Mee+YcPHnv6f+3dtnbKc5M33/ngx/7WTf3t89jTerdtHfiwB48BAAPhBEQAAOhImQYAgI6UaQAA6EiZBgCAjpRpAADoSJkGAICOlGkAAOhImQYAgI7GskxX1fKqumD16tXDjgIAwH5sLMt0a21la23FzMzMsKMAALAfG8syDQAAo0CZBgCAjpRpAADoSJkGAICOlGkAAOhImQYAgI6UaQAA6EiZBgCAjpRpAADoSJkGAICOlGkAAOhImQYAgI6UaQAA6EiZBgCAjiZ3dmdVHZPknCQ/nuQRSdYluTLJx5J8vLW2ZeAJt59reZLlS5cuHcbLAwBAkp0cma6q9yR5d5KNSf44yblJXpHk00nOTPKFqvqJvRFyW621la21FTMzM8N4eQAASLLzI9Nva61duZ3xK5N8qKqmkxw3mFgAADD6dnhkentFuqoOqarHzd2/sbV23SDDAQDAKNvlCYhVdXlVHVRVhyb5epL3VNWfDT4aAACMtn5W85hprd2d5PlJ3tNae2KS0wcbCwAARl8/ZXqyqo5K8oIklww4DwAAjI1+yvTvJbksyXWttSuq6lFJvj3YWAAAMPp2us50krTW/jbJ3261fX2Snx1kKAAAGAe7LNNVdUSS85Is2frxrbVfGlwsAAAYfbss00k+muTz6V2sZfNg4wAAwPjop0wvaq29fuBJAABgzPRzAuIlVfWsgScBAIAx00+Zfm16hXp9Vd0zd7t70MEAAGDU9bOax4F7IwgAAIybfuZMp6qek+Qn5jYvb625eAsAAPu9XU7zqKo/Sm+qx9Vzt9fOjQEAwH6tnyPTz0ry+NbaliSpqvcl+WqSNwwyGAAAjLp+TkBMkoO3+npmADkAAGDs9HNk+q1JvlpVn01S6c2dfuNAUwEAwBjoZzWPC6vq8iRPSq9Mv761dvOggwEAwKjb4TSPqjpp7s9TkxyVZFWSG5M8Ym4MAAD2azs7Mv1rSVYkedt27mtJnjaQRAAAMCZ2WKZbayvmvjyrtbZ+6/uqasGeDlJVJ6e3BN/hST7TWnvHnn4NAADYk/pZzeMf+xx7kKp6d1XdWlVXbjN+ZlVdW1XXVdUbkqS1dk1r7eVJXpBkWT/7BwCAYdrZnOmHV9UTkyysqidU1alzt59KsqjP/b83yZnb7HciyflJzkpySpJzq+qUufuek+QLST7zEL8PAADY63Y2Z/qZSV6S5Jgkf7bV+D1JfqufnbfWPldVS7YZPi3Jda2165Okqj6Y5LlJrm6tXZzk4qr6WJIP9PMaAAAwLDubM/2+JO+rqp9trf39HnzNo9NbFeQ+q5I8ee6I9/OTzE9y6Y6eXFUr0jsxMscdd9wejAUAAA9NP+tM/31V/UySH0qyYKvx3+v4mrX9l2mXJ7m8jzwXJLkgSZYtW9Y6ZgAAgN22yxMQq+qdSX4+yavTK8L/Mckjd+M1VyU5dqvtY5J8dzf2BwAAQ9HPah4/2lr7hSR3ttZ+N8lT8sAy/FBdkeTRVXV8VU0nOSfJxbuxPwAAGIp+yvR9a0yvrapHJJlNcnw/O6+qC5N8McmJVbWqql7WWtuU5FVJLktyTZKLWmtXPZTQVbW8qi5YvXr1Q3kaAADsUbucM51kZVUdnORPknwlvasfvqufnbfWzt3B+KXZyUmGfex3ZZKVy5YtO6/rPgAAYHfttExX1bz0rkZ4V5K/r6pLkixorTkkDADAfm+n0zxaa1uSvG2r7Q2KNAAA9PQzZ/qTVfWzVbW9Je0AAGC/1c+c6V9LsjjJ5qpal97yeK21dtBAk+1EVS1Psnzp0qXDigAAALs+Mt1aO7C1Nq+1NtVaO2hue2hFei7TytbaipmZmWHGAABgP9fPRVuqql5UVb8zt31sVZ02+GgAADDa+pkz/ZfpXajlP81tr0ly/sASAQDAmOhnzvSTW2unVtVXk6S1dufclQsBAGC/1s+R6dmqmkjvYi2pqiOSbBloql1wBUQAAEZBP2X67Uk+nOTIqvqDJF9I8ocDTbULTkAEAGAU7HKaR2vtb6rqn5M8Pb1l8c5urV0z8GQAADDi+pkznSTfTnL3fY+vquNaa/8+sFQAADAGdlmmq+rVSd6c5JYkmzN30ZYkjxtsNAAAGG39HJl+bZITW2u3DzoMAACMk35OQLwxyUgtm2E1DwAARkE/R6avT3J5VX0syYb7BltrfzawVLvQWluZZOWyZcvOG1YGAADop0z/+9xteu4GAACkv6XxfjdJqmpxa+3ewUcCAIDxsMs501X1lKq6Osk1c9v/oar+cuDJAABgxPVzAuKfJ3lmktuTpLX29SQ/McBMAAAwFvop02mt3bjN0OYBZAEAgLHSzwmIN1bVjyZpVTWd5DWZm/IBAAD7s36OTL88ySuTHJ1kVZLHz20PjXWmAQAYBbss062121prL2ytPay1dmRr7UXDvhpia21la23FzMzMMGMAALCf2+U0j6p6+3aGVyf5cmvto3s+EgAAjId+pnksSG9qx7fnbo9LcmiSl1XVnw8sGQAAjLh+TkBcmuRprbVNSVJV70jyySTPSPKNAWYDAICR1s+R6aOTLN5qe3GSR7TWNifZMJBUAAAwBvo5Mv3fknytqi5PUuldsOUPq2pxkk8PMBsAAIy0XZbp1tpfVdWlSU5Lr0z/Vmvtu3N3/5dBhgMAgFG2wzJdVUtaazckSWvte0k+us39leTo1tqqgSYEAIARtbMj039SVfPSK9H/nOT76a3ssTTJTyd5epI3p3chl72qqpYnWb506dK9/dIAAHC/HZ6A2Fr7j0l+J8mJSc5P8vn0ivUvJ7k2vRU+PrU3Qm4nm4u2AAAwdDudM91auzrJb++lLAAAMFb6WRoPAADYDmUaAAA6UqYBAKCjfi7akqp6TnoXa0mS/9NaWzm4SAAAMB52eWS6qt6a5LVJrp67vWZuDAAA9mv9HJn+mSSPb61tSZKqel+SryZ54yCDAQDAqOt3zvTBW31tcWcAAEh/R6bfmuSrVfXZJJXe3GlHpQEA2O/tsky31i6sqsuTPCm9Mv361trNgw62My4nDgDAKOjnBMRTkxyVZFWSG5M8oqpOqKq+VgIZBJcTBwBgFPRTiP8yyalJ/iW9I9OPnfv6sKp6eWvtkwPMBwAAI6ufExBvSPKE1tqy1toTkzwhyZVJTk/y3waYDQAARlo/Zfqk1tpV92201q5Or1xfP7hYAAAw+vqZ5nFtVb0jyQfntn8+ybeqan6S2YElAwCAEdfPkemXJLkuya8meV2S6+fGZpP89IByAQDAyOtnabx1Sd42d9vWmj2eCAAAxsQuy3RVPTq9C7eckmTBfeOttUcNMBcAAIy8fqZ5vCfJO5JsSm9ax18nef8gQwEAwDjop0wvbK19Jkm11r7TWntLkqcNNhYAAIy+flbzWF9V85J8u6peleSmJEcONhYAAIy+fo5M/2qSRUlek+SJSV6U5BcHmAkAAMZCP6t5XDH35ZokLx1sHAAAGB/9HJkGAAC2Q5kGAICOxrJMV9Xyqrpg9erVw44CAMB+bJdluqoeU1Wfqaor57YfV1VvGny0HWutrWytrZiZmRlmDAAA9nP9HJl+V5I3JplNktbavyQ5Z5ChAABgHPRTphe11r60zdimQYQBAIBx0k+Zvq2qTkjSkqSqfi7J9waaCgAAxkA/V0B8ZZILkpxUVTcl+bf0LtwCAAD7tX4u2nJ9ktOranGSea21ewYfCwAARt8uy3RVHZzkF5IsSTJZVUmS1tprBhkMAABGXT/TPC5N8k9JvpFky2DjAADA+OinTC9orf3awJMAAMCY6Wc1j/dX1XlVdVRVHXrfbeDJAABgxPVzZHpjkj9J8tuZWx5v7s9HDSoUAACMg37K9K8lWdpau23QYQAAYJz0M83jqiRrBx0EAADGTT9Hpjcn+VpVfTbJhvsGLY0HAMD+rp8y/ZG5GwAAsJV+roD4vr0RBAAAxs0Oy3RVXdRae0FVfSM/WMXjfq21xw00GQAAjLidHZl+7dyfz94bQQAAYNzscDWP1tr35r58RWvtO1vfkrxi78QDAIDR1c/SeM/YzthZezoIAACMm53Nmf7P6R2BflRV/ctWdx2Y5B8GHQwAAEbdzuZMfyDJx5O8Nckbthq/p7V2x0BTAQDAGNhhmW6trU6yOsm5eytMVZ2d5GeSHJnk/NbaJ/fWawMAwEPVz5zp3VJV766qW6vqym3Gz6yqa6vquqp6Q5K01j7SWjsvyUuS/PygswEAwO4YeJlO8t4kZ249UFUTSc5P70TGU5KcW1WnbPWQN83dDwAAI2vgZbq19rkk286xPi3Jda2161trG5N8MMlzq+ePk3y8tfaVQWcDAIDdsTeOTG/P0Ulu3Gp71dzYq5OcnuTnqurl23tiVa2oqi9X1Ze///3vDz4pAADswM5W8xik2s5Ya629Pcnbd/bE1toFSS5IkmXLlj3oMucAALC3DOvI9Kokx261fUyS7w4pCwAAdDKsMn1FkkdX1fFVNZ3knCQXDykLAAB0sjeWxrswyReTnFhVq6rqZa21TUleleSyJNckuai1dtVD2Ofyqrpg9erVgwkNAAB9GPic6dbadi/60lq7NMmlHfe5MsnKZcuWnbc72QAAYHcMa5oHAACMPWUaAAA6UqYBAKCjsSzTTkAEAGAUjGWZbq2tbK2tmJmZGXYUAAD2Y2NZpgEAYBQo0wAA0JEyDQAAHY1lmXYCIgAAo2Asy7QTEAEAGAVjWaYBAGAUKNMAANCRMg0AAB0p0wAA0NFYlmmreQAAMArGskxbzQMAgFEwlmUaAABGgTINAAAdKdMAANCRMg0AAB0p0wAA0JEyDQAAHY1lmbbONAAAo2Asy7R1pgEAGAVjWaYBAGAUKNMAANCRMg0AAB0p0wAA0JEyDQAAHSnTAADQkTINAAAdjWWZdtEWAABGwViWaRdtAQBgFIxlmQYAgFGgTAMAQEfKNAAAdKRMAwBAR8o0AAB0pEwDAEBHyjQAAHSkTAMAQEfKNAAAdDSWZdrlxAEAGAVjWaZdThwAgFEwlmUaAABGgTINAAAdKdMAANCRMg0AAB0p0wAA0JEyDQAAHSnTAADQkTINAAAdKdMAANCRMg0AAB0p0wAA0JEyDQAAHSnTAADQkTINAAAdjWWZrqrlVXXB6tWrhx0FAID92FiW6dbaytbaipmZmWFHAQBgPzaWZRoAAEaBMg0AAB0p0wAA0JEyDQAAHSnTAADQkTINAAAdKdMAANCRMg0AAB0p0wAA0JEyDQAAHSnTAADQkTINAAAdKdMAANCRMg0AAB0p0wAA0JEyDQAAHSnTAADQkTINAAAdKdMAANCRMg0AAB0p0wAA0NHIlOmqelRV/VVV/d2wswAAQD8GWqar6t1VdWtVXbnN+JlVdW1VXVdVb0iS1tr1rbWXDTIPAADsSYM+Mv3eJGduPVBVE0nOT3JWklOSnFtVpww4BwAA7HEDLdOttc8luWOb4dOSXDd3JHpjkg8mee4gcwAAwCAMY8700Ulu3Gp7VZKjq+qwqnpnkidU1Rt39OSqWlFVX66qL3//+98fdFYAANihySG8Zm1nrLXWbk/y8l09ubV2QZILkmTZsmVtD2cDAIC+DePI9Kokx261fUyS7w4hBwAA7JZhlOkrkjy6qo6vqukk5yS5eAg5AABgtwx6abwLk3wxyYlVtaqqXtZa25TkVUkuS3JNkotaa1c9xP0ur6oLVq9evedDAwBAnwY6Z7q1du4Oxi9Nculu7HdlkpXLli07r+s+AABgd43MFRABAGDcKNMAANCRMg0AAB2NZZl2AiIAAKNgLMt0a21la23FzMzMsKMAALAfG8syDQAAo0CZBgCAjpRpAADoaCzLtBMQAQAYBWNZpp2ACADAKBjLMg0AAKNAmQYAgI6UaQAA6EiZBgCAjsayTFvNAwCAUTCWZdpqHgAAjIKxLNMAADAKlGkAAOhImQYAgI6UaQAA6EiZBgCAjpRpAADoaCzLtHWmAQAYBWNZpq0zDQDAKBjLMg0AAKNAmQYAgI6UaQAA6EiZBgCAjpRpAADoSJkGAICOlGkAAOhoLMu0i7YAADAKxrJMu2gLAACjYCzLNAAAjAJlGgAAOlKmAQCgI2UaAAA6UqYBAKAjZRoAADpSpgEAoCNlGgAAOlKmAQCgo7Es0y4nDgDAKBjLMu1y4gAAjIKxLNMAADAKlGkAAOhImQYAgI6UaQAA6EiZBgCAjpRpAADoSJkGAICOlGkAAOhImQYAgI6UaQAA6EiZBgCAjpRpAADoSJkGAICOlGkAAOhoLMt0VS2vqgtWr1497CgAAOzHxrJMt9ZWttZWzMzMDDsKAAD7sbEs0wAAMAqUaQAA6EiZBgCAjpRpAADoSJkGAICOlGkAAOhImQYAgI6UaQAA6EiZBgCAjpRpAADoSJkGAICOlGkAAOhImQYAgI6UaQAA6EiZBgCAjpRpAADoSJkGAICOlGkAAOhImQYAgI6UaQAA6EiZBgCAjiaHHeA+VbU4yV8m2Zjk8tba3ww5EgAA7NRAj0xX1bur6taqunKb8TOr6tqquq6q3jA3/Pwkf9daOy/JcwaZCwAA9oRBT/N4b5Iztx6oqokk5yc5K8kpSc6tqlOSHJPkxrmHbR5wLgAA2G0DLdOttc8luWOb4dOSXNdau761tjHJB5M8N8mq9Ar1wHMBAMCeUK21wb5A1ZIkl7TWHju3/XNJzmyt/fLc9ouTPDnJ65P8RZL1Sb6woznTVbUiyYq5zQX37RcAAPa2YZyAWNsZa621e5O8dFdPbq1dkOSCPZ4KAAAeomFMp1iV5Nitto9J8t0h5AAAgN0yjDJ9RZJHV9XxVTWd5JwkFw8hBwAA7JZBL413YZIvJjmxqlZV1ctaa5uSvCrJZUmuSXJRa+2qQeZg+6qqVdXbttr+jap6y9zXb6mqm6rqa1vdDt7qsf9j7n4ni7LPqarN2/zbf8Pc+A1VdfhWj/upqrpk7uuXVNX35x7/zap63bDyA/ueqnre3M/tk+a2l1TVum3eq6a9F+19A50z3Vo7dwfjlya5dJCvTV82JHl+Vb21tXbbdu7/7621P912cK5APy+9pQx/IsnlA00Je9+61trjOzzvf7fWXlVVhyW5tqr+rrV24y6fBbBr5yb5Qnqf6L9lbuxft32vqqrEe9Fe5aji/m1TeidzPtTfWn86yZVJ3pHe/9zAVlprtye5LslRw84CjL+qOiDJjyV5WXplui/ei/YOZZrzk7ywqma2c9/rtvro6LNbjZ+b5MIkH07y7Kqa2htBYS9auM1Hpz//UJ5cVcclWZDkXwYTD9jPnJ3kE621byW5o6pOnRs/Yav3qfO3fZL3or1jGEvjMUJaa3dX1V8neU2Sddvc/aBpHnMnjT4ryetaa/dU1f+f5IwkH9srgWHv2NE0j+0tzL/12M9X1U8nOTHJea219YMIB+x3zk3y53Nff3Bu+/xsZ5rHHO9Fe5EyTdL7H/QrSd7Tx2PPTDKT5Btz87IWJVkbZZr9w+1JDkly3zkGh271dfKDeYpPSfKxqvp4a+3mvR0S2HfMzXt+WpLHVlVLMpHeL/F/uZOneS/ai0zzIK21O5JclN5crF05N8kvt9aWtNaWJDk+yRlVtWiAEWFUXJ7kxUlSVRNJXpTks9s+qLX2xSTvT/LavRkO2Cf9XJK/bq09cu5n77FJ/i2963TslPeivUOZ5j5vS3L4NmOv22be6ClJnpmtjkLPXbnyC0mW772oMHDbzpn+o7nx30+ytKq+nuSr6Z3Y8//uYB9/nOSlVXXgXsgL7LvOTe8cpa39fZLf6vP53osGrFrb3hRAAABgVxyZBgCAjpRpAADoSJkGAICOlGkAAOhImQYAgI6UaQAA6EiZBgCAjpRpAADo6P8CKqcz4ntDDbQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 864x648 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Convert the demographic object to a demes graph.\n",
    "test_graph = msprime.Demography.to_demes(test_model)\n",
    "# Visualize the demes graph.\n",
    "fig, ax = plt.subplots(figsize=(12,9))\n",
    "demesdraw.tubes(test_graph, ax=ax, log_time=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3af8f613-588b-4b7e-8637-2c7fc0a3cc56",
   "metadata": {},
   "source": [
    "## Example Simulation\n",
    "Here I will walk through a simulation using the `iua_human_model()` function we defined earlier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "adf5b9c6-8bb3-4c94-b348-f17a0e28ab04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>27827</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>20000000.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>generations</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>5</td></tr>\n",
       "                      <tr><td>Total Size</td><td>35.5 MiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>79693</td>\n",
       "                      <td>2.4 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>5</td>\n",
       "                      <td>164 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>35029</td>\n",
       "                      <td>1.5 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>516940</td>\n",
       "                      <td>18.2 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>16540</td>\n",
       "                      <td>452.3 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>7</td>\n",
       "                      <td>456 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>2</td>\n",
       "                      <td>3.9 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>516940</td>\n",
       "                      <td>12.3 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fc343764f10>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Simulate a tree sequence.\n",
    "ex_ts_f_03 = msprime.sim_ancestry(\n",
    "    samples=[\n",
    "        msprime.SampleSet(2, ploidy=1, population='AFR'), # Sample two haplotypes from AFR for our HMM.\n",
    "        msprime.SampleSet(1, ploidy=1, population='EUR'),\n",
    "        msprime.SampleSet(1, ploidy=1, population='NEA'),\n",
    "        msprime.SampleSet(1, ploidy=1, population='OUT'),\n",
    "    ],\n",
    "    demography=iua_human_model(0.03), # Rate of introgression at 0.03.\n",
    "    sequence_length=20_000_000, # Generate 20 Mb haplotypes per samples.\n",
    "    recombination_rate=1e-8,\n",
    "    record_migrations=True, # Need this to keep track of what segments are introgressed.\n",
    "    discrete_genome=False,\n",
    "    random_seed=42,\n",
    ")\n",
    "# Overlay mutations.\n",
    "ex_mts_f_03 = msprime.sim_mutations(\n",
    "    tree_sequence=ex_ts_f_03, rate=1.5e-8,\n",
    "    model='jc69', random_seed=42,\n",
    "    discrete_genome=False,\n",
    ")\n",
    "# View the mutated tree-sequence summary.\n",
    "ex_mts_f_03"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ceaa9e8b-9f39-4245-8cfb-ab73f4def1f7",
   "metadata": {},
   "source": [
    "Ok so now we will look at the genotype matrix, where the first two indicies (i.e., 0 and 1) are the two `AFR` haplotypes, the third index (i.e., 2) is the `EUR` haplotype, the fourth index (i.e., 3) is the `NEA` haplotype, and the last index (i.e., -1) is the `OUT` haplotype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0a5c1b56-00d3-4395-a157-40b591519fea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1 1 1 1 0]\n",
      " [1 1 1 1 0]\n",
      " [0 0 0 0 1]\n",
      " ...\n",
      " [1 1 1 1 0]\n",
      " [0 0 0 0 1]\n",
      " [0 0 0 0 1]]\n",
      "There are 516940 variants between 5 unpolarized haplotypes of 20 Mb\n"
     ]
    }
   ],
   "source": [
    "# View the genotype matrix.\n",
    "ex_raw_genotype_matrix = ex_mts_f_03.genotype_matrix()\n",
    "print(ex_raw_genotype_matrix)\n",
    "print('There are {0} variants between 5 unpolarized haplotypes of 20 Mb'.format(ex_raw_genotype_matrix.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d92baba9-f4e9-451b-98c5-5321b2c02ba1",
   "metadata": {},
   "source": [
    "Next lets just run another sanity check to make sure there are that many positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "711e01ce-229c-475d-9f70-b4770dec5300",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2.15364873e+01 3.33508083e+01 9.18967648e+01 ... 1.99999002e+07\n",
      " 1.99999936e+07 1.99999967e+07]\n",
      "There are 516940 variants between 5 unpolarized haplotypes of 20 Mb\n"
     ]
    }
   ],
   "source": [
    "# Grab the positions of the variable sites.\n",
    "ex_variable_positions = ex_mts_f_03.tables.sites.position\n",
    "print(ex_variable_positions)\n",
    "print('There are {0} variants between 5 unpolarized haplotypes of 20 Mb'.format(ex_variable_positions.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf4e66bd-b87c-4791-94af-af0e38cc26db",
   "metadata": {},
   "source": [
    "For our purposes we will need the polarize our haplotypes such that our outgroup haplotype is always 0, but no need to fear I will include a function on how to do so."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dcbae7c4-533c-4b06-8989-00f970a044d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1 1 1 1 0]\n",
      " [1 1 1 1 0]\n",
      " [1 1 1 1 0]\n",
      " ...\n",
      " [1 1 1 1 0]\n",
      " [1 1 1 1 0]\n",
      " [1 1 1 1 0]]\n"
     ]
    }
   ],
   "source": [
    "# Define a function to polarize a genotype matrix.\n",
    "def polarize_genotype_matrix(\n",
    "    genotype_matrix,\n",
    "):\n",
    "    \"\"\"\n",
    "    ###########################################################################\n",
    "    INPUT\n",
    "        genotype_matrix: Simulated genotypes from msprime.\n",
    "    ---------------------------------------------------------------------------\n",
    "    OUTPUT: Polarized genotype matrix where the outgroup always harbors the\n",
    "            ancestral allele.\n",
    "    ###########################################################################\n",
    "    \"\"\"\n",
    "    # Polarize the genotype matrix such that the outgroup taxon is always 0.\n",
    "    polarize = lambda x: abs(x - 1)\n",
    "    polarized_genotype_list = [polarize(site) if site[-1] == 1 else site for site in genotype_matrix]\n",
    "    polarized_genotype_matrix = np.asarray(polarized_genotype_list)\n",
    "    return polarized_genotype_matrix\n",
    "# Polarize the example genotype matrix.\n",
    "ex_polarized_genotype_matrix = polarize_genotype_matrix(ex_raw_genotype_matrix)\n",
    "print(ex_polarized_genotype_matrix)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cec478c5-3bb9-465a-b697-156309bff882",
   "metadata": {},
   "source": [
    "So now we need to find the start and stop positions of the introgressed segments, to do so we will parse the migrations table and extract the left (start) and right (stop) positions for all introgression treesi.e., the `dest` is from the `NEA` population (2) which occured 4000 generations ago (the time of introgression)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "27cca3d8-1cdc-4ef5-873b-3a95f1dc7454",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start positions of introgressed regions: [8140010.720961671, 8386088.554532683, 1448351.361984973, 2568334.2587925983, 233591.6468729049, 15527390.13840136, 4253073.412968563, 10434970.242597777, 16731242.807630714, 16967605.690655127, 19898089.55145829, 19426282.58320637, 11761718.294151794, 17951326.091634862, 260518.3202967598, 437440.6151562037, 17634818.185162324, 19572106.19887641]\n",
      "Stop positions of introgressed regions: [8140010.720961671, 8386088.554532683, 1448351.361984973, 2568334.2587925983, 233591.6468729049, 15527390.13840136, 4253073.412968563, 10434970.242597777, 16731242.807630714, 16967605.690655127, 19898089.55145829, 19426282.58320637, 11761718.294151794, 17951326.091634862, 260518.3202967598, 437440.6151562037, 17634818.185162324, 19572106.19887641]\n"
     ]
    }
   ],
   "source": [
    "# Intilaize the start and stop lists.\n",
    "ex_starts = []\n",
    "ex_stops = []\n",
    "# Parse through all migration events and record the start and stop positions for introgressed regions.\n",
    "for migration in ex_mts_f_03.migrations():\n",
    "    if (migration.dest == 2) and (migration.time == float(4000)):\n",
    "        ex_starts.append(migration.left)\n",
    "        ex_stops.append(migration.right)\n",
    "# Print a summary.\n",
    "print('Start positions of introgressed regions: {0}'.format(ex_starts))\n",
    "print('Stop positions of introgressed regions: {0}'.format(ex_starts))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a90c8af-a282-495c-92c7-f5ed8a371ae3",
   "metadata": {},
   "source": [
    "Now lets run one last sanity check by checking the actual migration table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "09c65dd1-2c8c-48eb-957c-10bfddd75549",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 18 start and 18 stop positions for introgressed regions.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "                <style scoped=\"\">\n",
       "                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}\n",
       "                    .tskit-table tbody tr th {vertical-align: top;}\n",
       "                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "                    .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "                </style>\n",
       "                <table border=\"1\" class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                        <tr>\n",
       "                            <th>id</th><th>left</th><th>right</th><th>node</th><th>source</th><th>dest</th><th>time</th><th>metadata</th>\n",
       "                        </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                        <tr><td>0</td><td>8140010.72096167</td><td>8142714.79592562</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>1</td><td>8386088.55453268</td><td>8400125.26909963</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>2</td><td>1448351.36198497</td><td>1463190.82120507</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>3</td><td>2568334.25879260</td><td>2689267.70426328</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>4</td><td>233591.64687290</td><td>260518.32029676</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>5</td><td>15527390.13840136</td><td>15561168.45830450</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>6</td><td>4253073.41296856</td><td>4259518.31279358</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>7</td><td>10434970.24259778</td><td>10463509.05945915</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>8</td><td>16731242.80763071</td><td>16803926.48858202</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>9</td><td>16967605.69065513</td><td>16974004.34731545</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>10</td><td>19898089.55145829</td><td>20000000.00000000</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>11</td><td>19426282.58320637</td><td>19444101.47957518</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>12</td><td>11761718.29415179</td><td>11785896.75502232</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>13</td><td>17951326.09163486</td><td>17998189.82477053</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>14</td><td>260518.32029676</td><td>316360.76562452</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>15</td><td>437440.61515620</td><td>441491.44720614</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>16</td><td>17634818.18516232</td><td>17653113.55657366</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>17</td><td>19572106.19887641</td><td>19573065.90233464</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>18</td><td>7472891.60870501</td><td>7483991.76513148</td><td>125</td><td>0</td><td>4</td><td>16000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>19</td><td>9538380.12771882</td><td>9541956.30041936</td><td>6</td><td>0</td><td>4</td><td>16000.00000000</td><td>b''</td></tr>\n",
       "<tr><td style=\"text-align: center;\" colspan=\"109\"><em>34989 rows skipped (tskit.set_print_options)</em></td></tr>\n",
       "<tr><td>35009</td><td>1762277.80828471</td><td>1769539.01518682</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35010</td><td>0.00000000</td><td>14401.74042944</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35011</td><td>3589169.51492844</td><td>3590582.86466586</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35012</td><td>3590675.67586210</td><td>3591039.61966448</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35013</td><td>3592382.33127358</td><td>3593157.37192324</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35014</td><td>3935835.26352424</td><td>3937660.26867319</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35015</td><td>3946585.88510351</td><td>3949807.81592103</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35016</td><td>5325917.72369040</td><td>5329026.47467290</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35017</td><td>5340126.71860600</td><td>5342350.37809845</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35018</td><td>5342762.54780446</td><td>5361045.59394739</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35019</td><td>5362911.01319668</td><td>5364872.91527985</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35020</td><td>5370615.24292468</td><td>5372031.42976222</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35021</td><td>5379046.59458601</td><td>5396055.65909495</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35022</td><td>5423914.82366783</td><td>5424123.98540465</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35023</td><td>5424641.61956398</td><td>5434592.69870755</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35024</td><td>5436043.37100211</td><td>5453084.44821858</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35025</td><td>5459466.63908824</td><td>5462737.49352673</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35026</td><td>5469545.04530156</td><td>5471702.36669926</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35027</td><td>5493817.52661629</td><td>5494718.14481958</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35028</td><td>7087710.48912274</td><td>7101304.14260444</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "\n",
       "                    </tbody>\n",
       "                </table>\n",
       "            </div>\n",
       "        "
      ],
      "text/plain": [
       "<tskit.tables.MigrationTable at 0x7fc342f9ba90>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Print the number of start and stop positions for introgressed regions.\n",
    "print('There are {0} start and {1} stop positions for introgressed regions.'.format(len(ex_starts), len(ex_stops)))\n",
    "# Print out the migration table.\n",
    "ex_mts_f_03.tables.migrations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8344b13d-11fa-4b71-8822-c147849b6048",
   "metadata": {},
   "source": [
    "Since all of our sanity checks look good lets generate this data!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59b402ab-896b-4268-80be-0b067176508c",
   "metadata": {},
   "source": [
    "## Generating Simulated Data Sanity Checks\n",
    "First I will just come up with a toy example, do a sanity check, and then actually explain everything."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7acadaa3-9b2b-41ad-926a-ffddc5cddd4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished simulating rep 1!\n"
     ]
    }
   ],
   "source": [
    "# For 1 replicate.\n",
    "for rep in list(range(1, 2)):\n",
    "    # Simulate a tree sequence.\n",
    "    ts_f_03 = msprime.sim_ancestry(\n",
    "        samples=[\n",
    "            msprime.SampleSet(2, ploidy=1, population='AFR'), # Sample two haplotypes from AFR for our HMM.\n",
    "            msprime.SampleSet(1, ploidy=1, population='EUR'),\n",
    "            msprime.SampleSet(1, ploidy=1, population='NEA'),\n",
    "            msprime.SampleSet(1, ploidy=1, population='OUT'),\n",
    "        ],\n",
    "        demography=iua_human_model(0.03), # Rate of introgression at 0.03.\n",
    "        sequence_length=20_000_000, # Generate 20 Mb haplotypes per samples.\n",
    "        recombination_rate=1e-8,\n",
    "        record_migrations=True, # Need this to keep track of what segments are introgressed.\n",
    "        discrete_genome=False,\n",
    "        random_seed=rep,\n",
    "    )\n",
    "    # Overlay mutations.\n",
    "    mts_f_03 = msprime.sim_mutations(\n",
    "        tree_sequence=ex_ts_f_03, rate=1.5e-8,\n",
    "        model='jc69', random_seed=rep,\n",
    "        discrete_genome=False,\n",
    "    )\n",
    "    # Extract the genotype matrix.\n",
    "    unpolarized_genotype_matrix = mts_f_03.genotype_matrix()\n",
    "    # Polarize the genotype matrix.\n",
    "    polarized_genotype_matrix = polarize_genotype_matrix(unpolarized_genotype_matrix)\n",
    "    # Save the genotype matrix.\n",
    "    np.savetxt(\n",
    "        './cs282_sim_data/rep_id_{0}_polarized_geno_mat.csv.gz'.format(rep),\n",
    "        polarized_genotype_matrix,\n",
    "        fmt='%d',\n",
    "        delimiter=',',\n",
    "        )\n",
    "    # Extract the variable positions.\n",
    "    variable_positions = mts_f_03.tables.sites.position\n",
    "    # Save the variable positions.\n",
    "    np.savetxt(\n",
    "        './cs282_sim_data/rep_id_{0}_var_pos.csv.gz'.format(rep),\n",
    "        [variable_positions],\n",
    "        fmt='%1.15f',\n",
    "        delimiter=',',\n",
    "        newline='\\n',\n",
    "    )\n",
    "    # Extract the start and stop positions for introgressed regions.\n",
    "    intro_starts = []\n",
    "    intro_stops = []\n",
    "    # Parse through all migration events and record the start and stop positions for introgressed regions.\n",
    "    for migration in mts_f_03.migrations():\n",
    "        if (migration.dest == 2) and (migration.time == float(4000)):\n",
    "            intro_starts.append(migration.left)\n",
    "            intro_stops.append(migration.right)\n",
    "    # Construct and output a pandas dataframe. \n",
    "    intro_regions_df = pd.DataFrame({\n",
    "        'start': intro_starts,\n",
    "        'stop': intro_stops,\n",
    "    })\n",
    "    # Save the pandas dataframe.\n",
    "    intro_regions_df.to_csv(\n",
    "        path_or_buf='./cs282_sim_data/rep_id_{0}_intro_pos.csv.gz'.format(rep),\n",
    "        sep=',',\n",
    "        float_format='%1.15f',\n",
    "        index=False,\n",
    "    )\n",
    "    # Save the tree-sequence just for good measures.\n",
    "    mts_f_03.dump('./cs282_sim_data/rep_id_{0}_mut_tree_seq.ts'.format(rep))\n",
    "    # Progress Checker\n",
    "    print('Finished simulating rep {0}!'.format(rep))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "de0d014c-c111-42d9-9743-3ee795e36fb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished simulating rep 1!\n"
     ]
    }
   ],
   "source": [
    "# For 1 replicate.\n",
    "for rep in list(range(1, 2)):\n",
    "    # Simulate a tree sequence.\n",
    "    ts_f_03 = msprime.sim_ancestry(\n",
    "        samples=[\n",
    "            msprime.SampleSet(2, ploidy=1, population='AFR'), # Sample two haplotypes from AFR for our HMM.\n",
    "            msprime.SampleSet(1, ploidy=1, population='EUR'),\n",
    "            msprime.SampleSet(1, ploidy=1, population='NEA'),\n",
    "        ],\n",
    "        demography=iua_human_model(0.03), # Rate of introgression at 0.03.\n",
    "        sequence_length=20_000_000, # Generate 20 Mb haplotypes per samples.\n",
    "        recombination_rate=1e-8,\n",
    "        record_migrations=True, # Need this to keep track of what segments are introgressed.\n",
    "        discrete_genome=False,\n",
    "        random_seed=rep,\n",
    "    )\n",
    "    # Overlay mutations.\n",
    "    mts_f_03 = msprime.sim_mutations(\n",
    "        tree_sequence=ex_ts_f_03, rate=1.5e-8,\n",
    "        model='jc69', random_seed=rep,\n",
    "        discrete_genome=False,\n",
    "    )\n",
    "    # Extract the genotype matrix.\n",
    "    genotype_matrix = mts_f_03.genotype_matrix()\n",
    "    # Save the genotype matrix.\n",
    "    np.savetxt(\n",
    "        './cs282_sim_data/three_pop_{0}_polarized_geno_mat.csv.gz'.format(rep),\n",
    "        genotype_matrix,\n",
    "        fmt='%d',\n",
    "        delimiter=',',\n",
    "        )\n",
    "    # Extract the variable positions.\n",
    "    variable_positions = mts_f_03.tables.sites.position\n",
    "    # Save the variable positions.\n",
    "    np.savetxt(\n",
    "        './cs282_sim_data/three_pop_{0}_var_pos.csv.gz'.format(rep),\n",
    "        [variable_positions],\n",
    "        fmt='%1.15f',\n",
    "        delimiter=',',\n",
    "        newline='\\n',\n",
    "    )\n",
    "    # Extract the start and stop positions for introgressed regions.\n",
    "    intro_starts = []\n",
    "    intro_stops = []\n",
    "    # Parse through all migration events and record the start and stop positions for introgressed regions.\n",
    "    for migration in mts_f_03.migrations():\n",
    "        if (migration.dest == 2) and (migration.time == float(4000)):\n",
    "            intro_starts.append(migration.left)\n",
    "            intro_stops.append(migration.right)\n",
    "    # Construct and output a pandas dataframe. \n",
    "    intro_regions_df = pd.DataFrame({\n",
    "        'start': intro_starts,\n",
    "        'stop': intro_stops,\n",
    "    })\n",
    "    # Save the pandas dataframe.\n",
    "    intro_regions_df.to_csv(\n",
    "        path_or_buf='./cs282_sim_data/three_pop_{0}_intro_pos.csv.gz'.format(rep),\n",
    "        sep=',',\n",
    "        float_format='%1.15f',\n",
    "        index=False,\n",
    "    )\n",
    "    # Save the tree-sequence just for good measures.\n",
    "    mts_f_03.dump('./cs282_sim_data/three_pop_{0}_mut_tree_seq.ts'.format(rep))\n",
    "    # Progress Checker\n",
    "    print('Finished simulating rep {0}!'.format(rep))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62a688be-a330-483e-9855-078d7e47cc22",
   "metadata": {},
   "source": [
    "Cool cool cool, now lets just run one last sanity check."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "15a9796b-4945-44c2-84fa-ab631b20bc62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>27827</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>20000000.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>generations</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>5</td></tr>\n",
       "                      <tr><td>Total Size</td><td>35.6 MiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">No Metadata</td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>79693</td>\n",
       "                      <td>2.4 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>5</td>\n",
       "                      <td>164 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>35029</td>\n",
       "                      <td>1.5 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>517909</td>\n",
       "                      <td>18.3 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>16540</td>\n",
       "                      <td>452.3 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>7</td>\n",
       "                      <td>456 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>2</td>\n",
       "                      <td>3.9 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>517909</td>\n",
       "                      <td>12.3 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fc342e9ed60>"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the mutated tree sequence.\n",
    "rep_id_1_mts = tskit.load('./cs282_sim_data/rep_id_1_mut_tree_seq.ts')\n",
    "# Load the variable positions.\n",
    "rep_id_1_var_pos = np.loadtxt('./cs282_sim_data/rep_id_1_var_pos.csv.gz', delimiter=',')\n",
    "# Load the genotype matrix.\n",
    "rep_id_1_polarized_geno_mat = np.loadtxt('./cs282_sim_data/rep_id_1_polarized_geno_mat.csv.gz', dtype=int, delimiter=',')\n",
    "# Load the introgressed region dataframe.\n",
    "rep_id_1_intro_pos_df = pd.read_csv('./cs282_sim_data/rep_id_1_intro_pos.csv.gz', float_precision='round_trip')\n",
    "# Inspect the tree-sequence summary.\n",
    "rep_id_1_mts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "66c23eb2-f7a2-4fc3-8c30-eb2badd8664e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "It is True that the variable positions are the same and True that the polarized genotype matrices are the same.\n"
     ]
    }
   ],
   "source": [
    "# Check to see if the variable positions and polarize genotype matricies are identical.\n",
    "var_pos_check = np.array_equal(rep_id_1_var_pos, rep_id_1_mts.tables.sites.position)\n",
    "geno_mat_check = np.array_equal(rep_id_1_polarized_geno_mat, polarize_genotype_matrix(rep_id_1_mts.genotype_matrix()))\n",
    "# Print the sanity check.\n",
    "print('It is {0} that the variable positions are the same and {1} that the polarized genotype matrices are the same.'.format(var_pos_check, geno_mat_check))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ac4aa45-ffb3-4d12-84d0-d749bc9b3f44",
   "metadata": {},
   "source": [
    "Lastly lets inspect that introgressed regions are correct."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1229cffe-3009-459c-ba0e-385d89da2764",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8.140011e+06</td>\n",
       "      <td>8.142715e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8.386089e+06</td>\n",
       "      <td>8.400125e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.448351e+06</td>\n",
       "      <td>1.463191e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.568334e+06</td>\n",
       "      <td>2.689268e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.335916e+05</td>\n",
       "      <td>2.605183e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1.552739e+07</td>\n",
       "      <td>1.556117e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4.253073e+06</td>\n",
       "      <td>4.259518e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.043497e+07</td>\n",
       "      <td>1.046351e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1.673124e+07</td>\n",
       "      <td>1.680393e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1.696761e+07</td>\n",
       "      <td>1.697400e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.989809e+07</td>\n",
       "      <td>2.000000e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1.942628e+07</td>\n",
       "      <td>1.944410e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1.176172e+07</td>\n",
       "      <td>1.178590e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1.795133e+07</td>\n",
       "      <td>1.799819e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2.605183e+05</td>\n",
       "      <td>3.163608e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>4.374406e+05</td>\n",
       "      <td>4.414914e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1.763482e+07</td>\n",
       "      <td>1.765311e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1.957211e+07</td>\n",
       "      <td>1.957307e+07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           start          stop\n",
       "0   8.140011e+06  8.142715e+06\n",
       "1   8.386089e+06  8.400125e+06\n",
       "2   1.448351e+06  1.463191e+06\n",
       "3   2.568334e+06  2.689268e+06\n",
       "4   2.335916e+05  2.605183e+05\n",
       "5   1.552739e+07  1.556117e+07\n",
       "6   4.253073e+06  4.259518e+06\n",
       "7   1.043497e+07  1.046351e+07\n",
       "8   1.673124e+07  1.680393e+07\n",
       "9   1.696761e+07  1.697400e+07\n",
       "10  1.989809e+07  2.000000e+07\n",
       "11  1.942628e+07  1.944410e+07\n",
       "12  1.176172e+07  1.178590e+07\n",
       "13  1.795133e+07  1.799819e+07\n",
       "14  2.605183e+05  3.163608e+05\n",
       "15  4.374406e+05  4.414914e+05\n",
       "16  1.763482e+07  1.765311e+07\n",
       "17  1.957211e+07  1.957307e+07"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Exported table.\n",
    "rep_id_1_intro_pos_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fd1747b6-27bc-43a3-8323-52a1b9b75839",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "                <style scoped=\"\">\n",
       "                    .tskit-table tbody tr th:only-of-type {vertical-align: middle;}\n",
       "                    .tskit-table tbody tr th {vertical-align: top;}\n",
       "                    .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "                    .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "                </style>\n",
       "                <table border=\"1\" class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                        <tr>\n",
       "                            <th>id</th><th>left</th><th>right</th><th>node</th><th>source</th><th>dest</th><th>time</th><th>metadata</th>\n",
       "                        </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                        <tr><td>0</td><td>8140010.72096167</td><td>8142714.79592562</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>1</td><td>8386088.55453268</td><td>8400125.26909963</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>2</td><td>1448351.36198497</td><td>1463190.82120507</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>3</td><td>2568334.25879260</td><td>2689267.70426328</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>4</td><td>233591.64687290</td><td>260518.32029676</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>5</td><td>15527390.13840136</td><td>15561168.45830450</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>6</td><td>4253073.41296856</td><td>4259518.31279358</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>7</td><td>10434970.24259778</td><td>10463509.05945915</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>8</td><td>16731242.80763071</td><td>16803926.48858202</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>9</td><td>16967605.69065513</td><td>16974004.34731545</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>10</td><td>19898089.55145829</td><td>20000000.00000000</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>11</td><td>19426282.58320637</td><td>19444101.47957518</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>12</td><td>11761718.29415179</td><td>11785896.75502232</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>13</td><td>17951326.09163486</td><td>17998189.82477053</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>14</td><td>260518.32029676</td><td>316360.76562452</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>15</td><td>437440.61515620</td><td>441491.44720614</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>16</td><td>17634818.18516232</td><td>17653113.55657366</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>17</td><td>19572106.19887641</td><td>19573065.90233464</td><td>2</td><td>1</td><td>2</td><td>4000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>18</td><td>7472891.60870501</td><td>7483991.76513148</td><td>125</td><td>0</td><td>4</td><td>16000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>19</td><td>9538380.12771882</td><td>9541956.30041936</td><td>6</td><td>0</td><td>4</td><td>16000.00000000</td><td>b''</td></tr>\n",
       "<tr><td style=\"text-align: center;\" colspan=\"109\"><em>34989 rows skipped (tskit.set_print_options)</em></td></tr>\n",
       "<tr><td>35009</td><td>1762277.80828471</td><td>1769539.01518682</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35010</td><td>0.00000000</td><td>14401.74042944</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35011</td><td>3589169.51492844</td><td>3590582.86466586</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35012</td><td>3590675.67586210</td><td>3591039.61966448</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35013</td><td>3592382.33127358</td><td>3593157.37192324</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35014</td><td>3935835.26352424</td><td>3937660.26867319</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35015</td><td>3946585.88510351</td><td>3949807.81592103</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35016</td><td>5325917.72369040</td><td>5329026.47467290</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35017</td><td>5340126.71860600</td><td>5342350.37809845</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35018</td><td>5342762.54780446</td><td>5361045.59394739</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35019</td><td>5362911.01319668</td><td>5364872.91527985</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35020</td><td>5370615.24292468</td><td>5372031.42976222</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35021</td><td>5379046.59458601</td><td>5396055.65909495</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35022</td><td>5423914.82366783</td><td>5424123.98540465</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35023</td><td>5424641.61956398</td><td>5434592.69870755</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35024</td><td>5436043.37100211</td><td>5453084.44821858</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35025</td><td>5459466.63908824</td><td>5462737.49352673</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35026</td><td>5469545.04530156</td><td>5471702.36669926</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35027</td><td>5493817.52661629</td><td>5494718.14481958</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "<tr><td>35028</td><td>7087710.48912274</td><td>7101304.14260444</td><td>4</td><td>3</td><td>6</td><td>800000.00000000</td><td>b''</td></tr>\n",
       "\n",
       "                    </tbody>\n",
       "                </table>\n",
       "            </div>\n",
       "        "
      ],
      "text/plain": [
       "<tskit.tables.MigrationTable at 0x7fc342f9bd90>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Full mutation table.\n",
    "rep_id_1_mts.tables.migrations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19e64740-6fee-4a60-ae80-865569d019dd",
   "metadata": {},
   "source": [
    "Great! Now that all sanity checks are complete lets generate this data for Brian!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3579df21-d5e4-4c23-9b9f-b34be13c2e91",
   "metadata": {},
   "source": [
    "## Data for Brian\n",
    "Here I simulate and export the following information to the `cs282_sim_data` directory:  \n",
    "`rep_id_{REP}_polarized_geno_mat.csv.gz` - polarized genotype matrix  \n",
    "`rep_id_{REP}_var_pos.csv.gz` - array of all variable positions  \n",
    "`rep_id_{REP}_intro_pos.csv.gz` - dataframe of start and stop positions for introgressed regions  \n",
    "`rep_id_{REP}_mut_tree_seq.ts` - the complete tree-sequence for the replicate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "18dd62bd-e972-4294-9db2-228ed0da2d08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished simulating rep 1!\n",
      "Finished simulating rep 2!\n",
      "Finished simulating rep 3!\n",
      "Finished simulating rep 4!\n",
      "Finished simulating rep 5!\n",
      "Finished simulating rep 6!\n",
      "Finished simulating rep 7!\n",
      "Finished simulating rep 8!\n",
      "Finished simulating rep 9!\n",
      "Finished simulating rep 10!\n",
      "Finished simulating rep 11!\n",
      "Finished simulating rep 12!\n",
      "Finished simulating rep 13!\n",
      "Finished simulating rep 14!\n",
      "Finished simulating rep 15!\n",
      "Finished simulating rep 16!\n",
      "Finished simulating rep 17!\n",
      "Finished simulating rep 18!\n",
      "Finished simulating rep 19!\n",
      "Finished simulating rep 20!\n",
      "Finished simulating rep 21!\n",
      "Finished simulating rep 22!\n",
      "Finished simulating rep 23!\n",
      "Finished simulating rep 24!\n",
      "Finished simulating rep 25!\n",
      "Finished simulating rep 26!\n",
      "Finished simulating rep 27!\n",
      "Finished simulating rep 28!\n",
      "Finished simulating rep 29!\n",
      "Finished simulating rep 30!\n",
      "Finished simulating rep 31!\n",
      "Finished simulating rep 32!\n",
      "Finished simulating rep 33!\n",
      "Finished simulating rep 34!\n",
      "Finished simulating rep 35!\n",
      "Finished simulating rep 36!\n",
      "Finished simulating rep 37!\n",
      "Finished simulating rep 38!\n",
      "Finished simulating rep 39!\n",
      "Finished simulating rep 40!\n",
      "Finished simulating rep 41!\n",
      "Finished simulating rep 42!\n",
      "Finished simulating rep 43!\n",
      "Finished simulating rep 44!\n",
      "Finished simulating rep 45!\n",
      "Finished simulating rep 46!\n",
      "Finished simulating rep 47!\n",
      "Finished simulating rep 48!\n",
      "Finished simulating rep 49!\n",
      "Finished simulating rep 50!\n",
      "Finished simulating rep 51!\n",
      "Finished simulating rep 52!\n",
      "Finished simulating rep 53!\n",
      "Finished simulating rep 54!\n",
      "Finished simulating rep 55!\n",
      "Finished simulating rep 56!\n",
      "Finished simulating rep 57!\n",
      "Finished simulating rep 58!\n",
      "Finished simulating rep 59!\n",
      "Finished simulating rep 60!\n",
      "Finished simulating rep 61!\n",
      "Finished simulating rep 62!\n",
      "Finished simulating rep 63!\n",
      "Finished simulating rep 64!\n",
      "Finished simulating rep 65!\n",
      "Finished simulating rep 66!\n",
      "Finished simulating rep 67!\n",
      "Finished simulating rep 68!\n",
      "Finished simulating rep 69!\n",
      "Finished simulating rep 70!\n",
      "Finished simulating rep 71!\n",
      "Finished simulating rep 72!\n",
      "Finished simulating rep 73!\n",
      "Finished simulating rep 74!\n",
      "Finished simulating rep 75!\n",
      "Finished simulating rep 76!\n",
      "Finished simulating rep 77!\n",
      "Finished simulating rep 78!\n",
      "Finished simulating rep 79!\n",
      "Finished simulating rep 80!\n",
      "Finished simulating rep 81!\n",
      "Finished simulating rep 82!\n",
      "Finished simulating rep 83!\n",
      "Finished simulating rep 84!\n",
      "Finished simulating rep 85!\n",
      "Finished simulating rep 86!\n",
      "Finished simulating rep 87!\n",
      "Finished simulating rep 88!\n",
      "Finished simulating rep 89!\n",
      "Finished simulating rep 90!\n",
      "Finished simulating rep 91!\n",
      "Finished simulating rep 92!\n",
      "Finished simulating rep 93!\n",
      "Finished simulating rep 94!\n",
      "Finished simulating rep 95!\n",
      "Finished simulating rep 96!\n",
      "Finished simulating rep 97!\n",
      "Finished simulating rep 98!\n",
      "Finished simulating rep 99!\n"
     ]
    }
   ],
   "source": [
    "# For 100 replicates.\n",
    "for rep in list(range(1, 101)):\n",
    "    # Simulate a tree sequence.\n",
    "    ts_f_03 = msprime.sim_ancestry(\n",
    "        samples=[\n",
    "            msprime.SampleSet(2, ploidy=1, population='AFR'), # Sample two haplotypes from AFR for our HMM.\n",
    "            msprime.SampleSet(1, ploidy=1, population='EUR'),\n",
    "            msprime.SampleSet(1, ploidy=1, population='NEA'),\n",
    "            msprime.SampleSet(1, ploidy=1, population='OUT'),\n",
    "        ],\n",
    "        demography=iua_human_model(0.03), # Rate of introgression at 0.03.\n",
    "        sequence_length=20_000_000, # Generate 20 Mb haplotypes per samples.\n",
    "        recombination_rate=1e-8,\n",
    "        record_migrations=True, # Need this to keep track of what segments are introgressed.\n",
    "        discrete_genome=False,\n",
    "        random_seed=rep,\n",
    "    )\n",
    "    # Overlay mutations.\n",
    "    mts_f_03 = msprime.sim_mutations(\n",
    "        tree_sequence=ex_ts_f_03, rate=1.5e-8,\n",
    "        model='jc69', random_seed=rep,\n",
    "        discrete_genome=False,\n",
    "    )\n",
    "    # Extract the genotype matrix.\n",
    "    unpolarized_genotype_matrix = mts_f_03.genotype_matrix()\n",
    "    # Polarize the genotype matrix.\n",
    "    polarized_genotype_matrix = polarize_genotype_matrix(unpolarized_genotype_matrix)\n",
    "    # Save the genotype matrix.\n",
    "    np.savetxt(\n",
    "        './cs282_sim_data/rep_id_{0}_polarized_geno_mat.csv.gz'.format(rep),\n",
    "        polarized_genotype_matrix,\n",
    "        fmt='%d',\n",
    "        delimiter=',',\n",
    "        )\n",
    "    # Extract the variable positions.\n",
    "    variable_positions = mts_f_03.tables.sites.position\n",
    "    # Save the variable positions.\n",
    "    np.savetxt(\n",
    "        './cs282_sim_data/rep_id_{0}_var_pos.csv.gz'.format(rep),\n",
    "        [variable_positions],\n",
    "        fmt='%1.15f',\n",
    "        delimiter=',',\n",
    "        newline='\\n',\n",
    "    )\n",
    "    # Extract the start and stop positions for introgressed regions.\n",
    "    intro_starts = []\n",
    "    intro_stops = []\n",
    "    # Parse through all migration events and record the start and stop positions for introgressed regions.\n",
    "    for migration in mts_f_03.migrations():\n",
    "        if (migration.dest == 2) and (migration.time == float(4000)):\n",
    "            intro_starts.append(migration.left)\n",
    "            intro_stops.append(migration.right)\n",
    "    # Construct and output a pandas dataframe. \n",
    "    intro_regions_df = pd.DataFrame({\n",
    "        'start': intro_starts,\n",
    "        'stop': intro_stops,\n",
    "    })\n",
    "    # Save the pandas dataframe.\n",
    "    intro_regions_df.to_csv(\n",
    "        path_or_buf='./cs282_sim_data/rep_id_{0}_intro_pos.csv.gz'.format(rep),\n",
    "        sep=',',\n",
    "        float_format='%1.15f',\n",
    "        index=False,\n",
    "    )\n",
    "    # Save the tree-sequence just for good measures.\n",
    "    mts_f_03.dump('./cs282_sim_data/rep_id_{0}_mut_tree_seq.ts'.format(rep))\n",
    "    # Progress Checker\n",
    "    print('Finished simulating rep {0}!'.format(rep))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9c043ff-7219-49c3-a51f-9bc0cca44c5c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
